# Интегрирование с помощью формул Гаусса

In [87]:
import math

Многочлен лежандра третьей степени:

$$P(x) = x \left(x^2 - \frac35 \right)$$
$$x_1 = -\sqrt{\dfrac{3}{5}}, \, x_2 = 0, \, x_3 = \sqrt{\dfrac{3}{5}}$$

Смасштабируем х на отрезок $[a, b]$.

$$y(x) = x \dfrac{b - a}2 + \dfrac{b + a}2$$

Возьмём $a = 3, b = 11$, $f(x) = \sinh(x)$


In [104]:
def y(x, a, b):
    return x * (b - a) / 2 + (b + a) / 2

#формула гаусса
def Gauss(f, a, b):
    value = 0
    x = [-math.sqrt(3/5), 0, math.sqrt(3/5)]
    c = [5 / 9, 8 / 9, 5 / 9]
    for i in range(3):
        value += c[i] * f(y(x[i], a, b))
    return value * (b - a) / 2

def Integrate(h, f, a, b):
    value = 0
    i = 0
    while(abs(a + h * i - b) > 1e-10):
        value += Gauss(f, a + h * i, a + h * (i + 1)) 
        i += 1
    return value

#реализация алгоритма
def solve(a, b, f, epsA = 1e-5, epsO = 1e-5):
    alpha, beta = a, b
    I = E = 0
    while alpha < beta:        
        h = beta - alpha
        I_h = Integrate(h, f, alpha, beta)
        I_hdiv2 = Integrate(h / 2, f, alpha, beta)
        I_hdiv4 = Integrate(h / 4, f, alpha, beta)                
#Погрешность методом Эйткена
        delta = ((I_hdiv2 - I_h)**2 / (2 * I_hdiv2 - I_h - I_hdiv4))
        
        if abs(delta) < max(epsA, epsO * abs(I)) * h / (b - a):
            alpha = beta
            beta = min(beta + h, b)
            I = I + I_hdiv2 + delta
            E = E + abs((I_hdiv2 - I_h)**2 / (2 * I_hdiv2 - I_h - I_hdiv4))

        if abs(delta) > max(epsA, epsO * abs(I)) * h / (b - a):
            beta = (alpha + beta) / 2

        elif abs(delta) < max(epsA, epsO * abs(I)) / 2 ** 3 * h / (b - a):
            beta = min(alpha + 2 * h, b)

    return I, E

In [105]:
a, b = 0, 2 * math.pi
f = lambda x: math.cos(x) * math.exp(x)
res = solve(a, b, f)

print("S(f) = %f, суммарная погрешность главного члена равна %f" % (res[0], res[1]))

S(f) = 267.245799, суммарная погрешность главного члена равна 0.000030


Как и ожидалось $$\int_0^{2\pi} e^x \cos x dx = e^{\pi} \sinh \pi \approx 267.25$$

In [106]:
math.exp(math.pi) * math.sinh(math.pi)

267.24582776238236